File: packages/python/onnx9000-tflite-exporter/src/onnx9000/tflite_exporter/flatbuffer/builder.py
import struct
from typing import Optional, Union


class FlatBufferBuilder:
    """Zero-dependency FlatBuffer Builder in Python."""

    def __init__(self, initial_size: int = 1024) -> None:
        """Initialize the FlatBufferBuilder."""
        self.bb = bytearray(initial_size)
        self.space = initial_size
        self.minalign = 1
        self.vtable: Optional[list[int]] = None
        self.object_start = 0
        self.vtables: list[int] = []

    def clear(self) -> None:
        """Clear the builder."""
        self.space = len(self.bb)
        self.minalign = 1
        self.vtable = None
        self.object_start = 0
        self.vtables = []

    def grow_buffer(self) -> None:
        """Double the size of the buffer."""
        old_length = len(self.bb)
        new_length = old_length * 2
        if new_length == 0:
            new_length = 1024
        new_bb = bytearray(new_length)
        new_bb[new_length - old_length :] = self.bb
        self.bb = new_bb
        self.space += new_length - old_length

    def prep(self, size: int, additional_bytes: int) -> None:
        """Prepare to write an element of `size` after `additional_bytes` have been written."""
        if size > self.minalign:
            self.minalign = size
        align_size = (~(len(self.bb) - self.space + additional_bytes) + 1) & (size - 1)
        while self.space < align_size + size + additional_bytes:
            self.grow_buffer()
        self.pad(align_size)

    def pad(self, bytes_to_pad: int) -> None:
        """Pad the buffer."""
        for _ in range(bytes_to_pad):
            self.space -= 1
            self.bb[self.space] = 0

    def place(self, x: int) -> None:
        """Place a 32-bit int."""
        self.space -= 4
        struct.pack_into("<i", self.bb, self.space, x)

    def place_int8(self, x: int) -> None:
        """Place an 8-bit int."""
        self.space -= 1
        x = x if x <= 127 else x - 256
        x = x if x >= -128 else x + 256
        struct.pack_into("<b", self.bb, self.space, x)

    def place_int16(self, x: int) -> None:
        """Place a 16-bit int."""
        self.space -= 2
        struct.pack_into("<h", self.bb, self.space, x)

    def place_int32(self, x: int) -> None:
        """Place a 32-bit int."""
        self.space -= 4
        struct.pack_into("<i", self.bb, self.space, x)

    def place_int64(self, x: int) -> None:
        """Place a 64-bit int."""
        self.space -= 8
        struct.pack_into("<q", self.bb, self.space, x)

    def place_float32(self, x: float) -> None:
        """Place a 32-bit float."""
        self.space -= 4
        struct.pack_into("<f", self.bb, self.space, x)

    def place_float64(self, x: float) -> None:
        """Place a 64-bit float."""
        self.space -= 8
        struct.pack_into("<d", self.bb, self.space, x)

    def add_int8(self, x: int) -> None:
        """Add an 8-bit int."""
        self.prep(1, 0)
        self.place_int8(x)

    def add_int16(self, x: int) -> None:
        """Add a 16-bit int."""
        self.prep(2, 0)
        self.place_int16(x)

    def add_int32(self, x: int) -> None:
        """Add a 32-bit int."""
        self.prep(4, 0)
        self.place_int32(x)

    def add_int64(self, x: int) -> None:
        """Add a 64-bit int."""
        self.prep(8, 0)
        self.place_int64(x)

    def add_float32(self, x: float) -> None:
        """Add a 32-bit float."""
        self.prep(4, 0)
        self.place_float32(x)

    def add_float64(self, x: float) -> None:
        """Add a 64-bit float."""
        self.prep(8, 0)
        self.place_float64(x)

    def add_offset(self, offset: int) -> None:
        """Add an offset."""
        self.prep(4, 0)
        self.place(self.offset() - offset + 4)

    def start_vector(self, elem_size: int, num_elems: int, alignment: int) -> None:
        """Start a vector."""
        self.prep(4, elem_size * num_elems)
        self.prep(alignment, elem_size * num_elems)

    def end_vector(self, num_elems: int) -> int:
        """End a vector."""
        self.place_int32(num_elems)
        return self.offset()

    def create_byte_vector(
        self, data: Union[bytes, bytearray, list[int]], alignment: int = 4
    ) -> int:
        """Create a byte vector with specific alignment."""
        self.start_vector(1, len(data), alignment)
        self.space -= len(data)
        self.bb[self.space : self.space + len(data)] = data
        return self.end_vector(len(data))

    def create_string(self, s: Union[str, bytes, bytearray]) -> int:
        """Create a string or binary buffer."""
        if isinstance(s, str):
            utf8 = s.encode("utf-8")
        else:
            utf8 = bytes(s)
        self.add_int8(0)
        self.start_vector(1, len(utf8), 1)
        self.space -= len(utf8)
        self.bb[self.space : self.space + len(utf8)] = utf8
        return self.end_vector(len(utf8))

    def start_object(self, numfields: int) -> None:
        """Start an object."""
        self.vtable = [0] * numfields
        self.object_start = self.offset()

    def add_field_int8(self, voffset: int, x: int, d: int) -> None:
        """Add an 8-bit int field."""
        if x != d:
            self.add_int8(x)
            self.slot(voffset)

    def add_field_int16(self, voffset: int, x: int, d: int) -> None:
        """Add a 16-bit int field."""
        if x != d:
            self.add_int16(x)
            self.slot(voffset)

    def add_field_int32(self, voffset: int, x: int, d: int) -> None:
        """Add a 32-bit int field."""
        if x != d:
            self.add_int32(x)
            self.slot(voffset)

    def add_field_int64(self, voffset: int, x: int, d: int) -> None:
        """Add a 64-bit int field."""
        if x != d:
            self.add_int64(x)
            self.slot(voffset)

    def add_field_float32(self, voffset: int, x: float, d: float) -> None:
        """Add a 32-bit float field."""
        if x != d:
            self.add_float32(x)
            self.slot(voffset)

    def add_field_float64(self, voffset: int, x: float, d: float) -> None:
        """Add a 64-bit float field."""
        if x != d:
            self.add_float64(x)
            self.slot(voffset)

    def add_field_offset(self, voffset: int, x: int, d: int) -> None:
        """Add an offset field."""
        if x != d:
            self.add_offset(x)
            self.slot(voffset)

    def slot(self, voffset: int) -> None:
        """Slot a field."""
        if self.vtable is not None:
            self.vtable[voffset] = self.offset()

    def end_object(self) -> int:
        """End an object."""
        if self.vtable is None:
            raise ValueError("FlatBufferBuilder: end_object called without start_object")
        self.add_int32(0)
        vtableloc = self.offset()

        # write vtable entries
        i = len(self.vtable) - 1
        while i >= 0 and self.vtable[i] == 0:
            i -= 1
        trimmed_size = i + 1

        for j in range(trimmed_size - 1, -1, -1):
            off = vtableloc - self.vtable[j] if self.vtable[j] else 0
            self.add_int16(off)

        standard_fields = 2
        self.add_int16(vtableloc - self.object_start)
        self.add_int16((trimmed_size + standard_fields) * 2)

        existing_vtable = 0
        vt1 = self.space
        vt1len = struct.unpack_from("<h", self.bb, vt1)[0]

        for vt2_offset in self.vtables:
            vt2 = len(self.bb) - vt2_offset
            vt2len = struct.unpack_from("<h", self.bb, vt2)[0]
            if vt1len == vt2len:
                match = True
                for k in range(2, vt1len, 2):
                    v1 = struct.unpack_from("<h", self.bb, vt1 + k)[0]
                    v2 = struct.unpack_from("<h", self.bb, vt2 + k)[0]
                    if v1 != v2:
                        match = False
                        break
                if match:
                    existing_vtable = vt2_offset
                    break

        if existing_vtable:
            self.space = len(self.bb) - vtableloc
            struct.pack_into("<i", self.bb, self.space, existing_vtable - vtableloc)
        else:
            self.vtables.append(self.offset())
            struct.pack_into("<i", self.bb, len(self.bb) - vtableloc, self.offset() - vtableloc)

        self.vtable = None
        return vtableloc

    def finish(self, root_table: int, identifier: Optional[str] = None) -> None:
        """Finish the buffer."""
        self.prep(self.minalign, 8 if identifier else 4)
        if identifier:
            al = 4
            self.prep(al, 4)
            for i in range(3, -1, -1):
                self.add_int8(ord(identifier[i]))
        self.add_offset(root_table)

    def offset(self) -> int:
        """Get the current offset."""
        return len(self.bb) - self.space

    def as_bytearray(self) -> bytearray:
        """Get the buffer as a bytearray."""
        return self.bb[self.space :]

File: packages/js/tflite-exporter/src/flatbuffer/builder.ts
/* eslint-disable */
/**
 * Zero-dependency FlatBuffer Builder in TypeScript.
 */
export class FlatBufferBuilder {
  private bb: Uint8Array;
  private space: number;
  private minalign: number = 1;
  private vtable: number[] | null = null;
  private objectStart: number = 0;
  private vtables: number[] = [];
  private view: DataView;

  constructor(initialSize: number = 1024) {
    this.bb = new Uint8Array(initialSize);
    this.view = new DataView(this.bb.buffer);
    this.space = initialSize;
  }

  public clear(): void {
    this.space = this.bb.length;
    this.minalign = 1;
    this.vtable = null;
    this.objectStart = 0;
    this.vtables = [];
  }

  private growBuffer(): void {
    const oldLength = this.bb.length;
    let newLength = oldLength * 2;
    if (newLength === 0) newLength = 1024;

    // 25. Support chunked writing for models exceeding JS ArrayBuffer limits (>2GB).
    // V8 ArrayBuffer max length is ~2GB (2,147,483,647 bytes on 32-bit architecture limits, up to 4GB on 64-bit Chrome).
    // Node.js max is typically ~2GB buffer size.
    const MAX_ARRAY_BUFFER_SIZE = 2147483647; // 2GB limit
    if (newLength > MAX_ARRAY_BUFFER_SIZE) {
      if (oldLength < MAX_ARRAY_BUFFER_SIZE) {
        newLength = MAX_ARRAY_BUFFER_SIZE; // Push exactly to the limit first
      } else {
        throw new Error(
          `[onnx2tf] Fatal: TFLite model exceeds absolute v8 ArrayBuffer maximum limits (2GB). Use chunked streaming writer or target Python native exporter for massive parameter isolation.`,
        );
      }
    }

    const newBb = new Uint8Array(newLength);
    newBb.set(this.bb, newLength - oldLength);
    this.bb = newBb;
    this.view = new DataView(this.bb.buffer);
    this.space += newLength - oldLength;
  }

  private prep(size: number, additionalBytes: number): void {
    if (size > this.minalign) {
      this.minalign = size;
    }
    const alignSize = (~(this.bb.length - this.space + additionalBytes) + 1) & (size - 1);
    while (this.space < alignSize + size + additionalBytes) {
      this.growBuffer();
    }
    this.pad(alignSize);
  }

  private pad(bytes: number): void {
    for (let i = 0; i < bytes; i++) {
      this.bb[--this.space] = 0;
    }
  }

  public place(x: number): void {
    this.space -= 4;
    this.view.setInt32(this.space, x, true);
  }

  public placeInt8(x: number): void {
    this.bb[--this.space] = x;
  }

  public placeInt16(x: number): void {
    this.space -= 2;
    this.view.setInt16(this.space, x, true);
  }

  public placeInt32(x: number): void {
    this.space -= 4;
    this.view.setInt32(this.space, x, true);
  }

  public placeInt64(x: bigint): void {
    this.space -= 8;
    this.view.setBigInt64(this.space, x, true);
  }

  public placeFloat32(x: number): void {
    this.space -= 4;
    this.view.setFloat32(this.space, x, true);
  }

  public placeFloat64(x: number): void {
    this.space -= 8;
    this.view.setFloat64(this.space, x, true);
  }

  public addInt8(x: number): void {
    this.prep(1, 0);
    this.placeInt8(x);
  }

  public addInt16(x: number): void {
    this.prep(2, 0);
    this.placeInt16(x);
  }

  public addInt32(x: number): void {
    this.prep(4, 0);
    this.placeInt32(x);
  }

  public addInt64(x: bigint): void {
    this.prep(8, 0);
    this.placeInt64(x);
  }

  public addFloat32(x: number): void {
    this.prep(4, 0);
    this.placeFloat32(x);
  }

  public addFloat64(x: number): void {
    this.prep(8, 0);
    this.placeFloat64(x);
  }

  public addOffset(offset: number): void {
    this.prep(4, 0);
    this.place(this.offset() - offset + 4);
  }

  public startVector(elemSize: number, numElems: number, alignment: number): void {
    this.prep(4, elemSize * numElems);
    this.prep(alignment, elemSize * numElems);
  }

  public endVector(numElems: number): number {
    this.placeInt32(numElems);
    return this.offset();
  }

  public createByteVector(data: Uint8Array | number[], alignment: number = 4): number {
    this.startVector(1, data.length, alignment);
    this.space -= data.length;
    this.bb.set(data, this.space);
    return this.endVector(data.length);
  }

  public createString(s: string | Uint8Array): number {
    let utf8: Uint8Array;
    if (typeof s === 'string') {
      utf8 = new TextEncoder().encode(s);
    } else {
      utf8 = s;
    }
    this.addInt8(0);
    this.startVector(1, utf8.length, 1);
    this.space -= utf8.length;
    this.bb.set(utf8, this.space);
    return this.endVector(utf8.length);
  }

  public startObject(numfields: number): void {
    this.vtable = new Array(numfields).fill(0);
    this.objectStart = this.offset();
  }

  public addFieldInt8(voffset: number, x: number, d: number): void {
    if (x !== d) {
      this.addInt8(x);
      this.slot(voffset);
    }
  }

  public addFieldInt16(voffset: number, x: number, d: number): void {
    if (x !== d) {
      this.addInt16(x);
      this.slot(voffset);
    }
  }

  public addFieldInt32(voffset: number, x: number, d: number): void {
    if (x !== d) {
      this.addInt32(x);
      this.slot(voffset);
    }
  }

  public addFieldInt64(voffset: number, x: bigint, d: bigint): void {
    if (x !== d) {
      this.addInt64(x);
      this.slot(voffset);
    }
  }

  public addFieldFloat32(voffset: number, x: number, d: number): void {
    if (x !== d) {
      this.addFloat32(x);
      this.slot(voffset);
    }
  }

  public addFieldFloat64(voffset: number, x: number, d: number): void {
    if (x !== d) {
      this.addFloat64(x);
      this.slot(voffset);
    }
  }

  public addFieldOffset(voffset: number, x: number, d: number): void {
    if (x !== d) {
      this.addOffset(x);
      this.slot(voffset);
    }
  }

  public slot(voffset: number): void {
    if (this.vtable !== null) {
      this.vtable[voffset] = this.offset();
    }
  }

  public endObject(): number {
    if (this.vtable === null) {
      throw new Error('FlatBufferBuilder: endObject called without startObject');
    }
    this.addInt32(0); // placeholder for vtable offset
    const vtableloc = this.offset();

    // write vtable entries
    let i = this.vtable.length - 1;
    for (; i >= 0 && this.vtable[i] === 0; i--) {}
    const trimmedSize = i + 1;

    for (let j = trimmedSize - 1; j >= 0; j--) {
      const off = this.vtable[j]! ? vtableloc - this.vtable[j]! : 0;
      this.addInt16(off);
    }

    const standardFields = 2; // vtable size, object size
    this.addInt16(vtableloc - this.objectStart);
    this.addInt16((trimmedSize + standardFields) * 2);

    let existingVtable = 0;
    const vt1 = this.space;
    const vt1len = this.view.getInt16(vt1, true);

    for (let j = 0; j < this.vtables.length; j++) {
      const vt2 = this.bb.length - this.vtables[j]!;
      const vt2len = this.view.getInt16(vt2, true);
      if (vt1len === vt2len) {
        let match = true;
        for (let k = 2; k < vt1len; k += 2) {
          if (this.view.getInt16(vt1 + k, true) !== this.view.getInt16(vt2 + k, true)) {
            match = false;
            break;
          }
        }
        if (match) {
          existingVtable = this.vtables[j]!;
          break;
        }
      }
    }

    if (existingVtable) {
      this.space = this.bb.length - vtableloc; // remove vtable from buffer, keeping the 4-byte offset placeholder
      this.view.setInt32(this.space, existingVtable - vtableloc, true);
    } else {
      this.vtables.push(this.offset());
      this.view.setInt32(this.bb.length - vtableloc, this.offset() - vtableloc, true);
    }

    this.vtable = null;
    return vtableloc;
  }

  public finish(rootTable: number, identifier?: string): void {
    this.prep(this.minalign, identifier ? 8 : 4);
    if (identifier) {
      const al = 4;
      this.prep(al, 4);
      for (let i = 3; i >= 0; i--) {
        this.addInt8(identifier.charCodeAt(i));
      }
    }
    this.addOffset(rootTable);
  }

  public offset(): number {
    return this.bb.length - this.space;
  }

  public asUint8Array(): Uint8Array {
    return this.bb.subarray(this.space, this.bb.length);
  }
}
